torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp

```diff
@@ -1428,11 +1520,23 @@ void ExprSegmentationSorter::sort() {
 }
 
 std::vector<Expr*> ExprSegmentationSorter::getExprs() const {
-  std::vector<Expr*> exprs;
+  // At this stage, there is no data dependency between different groups, so we
+  // can put their exprs interleavely. We choose to put scalar expressions at
+  // the beginning, so that indexing/predicates can reuse these computation.
+  std::vector<Expr*> scalar_exprs;
+  std::vector<Expr*> non_scalar_exprs;
   for (auto& group : groups_) {
-    exprs.insert(exprs.end(), group->exprs().begin(), group->exprs().end());
+    std::vector<Expr*>* active_exprs = &scalar_exprs;
+    for (auto expr : group->exprs()) {
+      if (!lower_utils::isScalarExpr(expr)) {
+        active_exprs = &non_scalar_exprs;
+      }
```

Comment:
Once a non-scalar expr is found, are all the remaining exprs also non-scalar? Or, there can still be scalar expressions, but since we can't move them to the front, we just leave there in the `non_scalar_exprs` list?

Reply:
It's the latter:
> there can still be scalar expressions, but since we can't move them to the front, we just leave there in the `non_scalar_exprs` list?

Reply:
Just renamed them into `scalar_exprs_without_tv_dep` and `remaining_exprs`